### PR TITLE
Fix `GDScriptCache::get_full_script` eating parsing errors because of early exit

### LINKED ISSUE #75545: No runtime errors when running a script with a syntax error (Unexpected "identifier" in class body)
### Godot version

 Godot Engine v4.0.1.stable.official.cacf49999

### System information

Windows 10,  Vulkan API 1.3.205 - Forward+ - Using Vulkan Device #0: NVIDIA - NVIDIA GeFor ce RTX 3080

### Issue description

Creates coding in an external text editor.

Run the script as follows.

godot_console.exe --ext gd

### Steps to reproduce

no code

### Minimal reproduction project

[test_script.zip](https://github.com/godotengine/godot/files/11128000/test_script.zip)

### PR DESCRIPTION
**Pull request by mattbork**
Historically and up through Godot 4.0-beta4, parse errors in a malformed GDScript will be printed to output and reported to the debugger at runtime to interrupt execution. In 4.0-beta5 and later, parse errors for most scripts are no longer reported, and the debugger does not stop execution. (I say most because scripts used as named classes or preloads, e.g., still fail to resolve or load as a resource). Up through 4.1.1-stable, there is no indication at all in the editor that a script has failed to be run because of a parsing error, which is potentially quite confusing. In 4.1.2-stable, the addition of https://github.com/godotengine/godot/pull/78540/ at least prints an error when `ResourceFormatLoaderGDScript::load` fails, which a user can see in the debug panel. However, it is easily missed as execution is still not interrupted, and it does not show the actual parsing error anyway.

The root cause is a change in the responsiblity and logic of `GDScriptCache::get_shallow_script` introduced in https://github.com/godotengine/godot/pull/68374. Prior to that, `get_shallow_script` only loaded and cached the source code of a script. `GDScriptCache::get_full_script` would exit early if `get_shallow_script` failed because there would have been no script to call `GDScript::reload` on (which does the actual parsing and error reporting). Since the changes in the above PR, however, `get_shallow_script` now also parses the source code and passes along any errors. But `get_full_script` still exits early if `get_shallow_script` fails for any reason. This means that a script that loads sucessfully but hits an error during parsing will not make it to the `reload` call. Thus, the errors are never printed and never reported to the debugger.

The simplest fix is to exit from `get_full_script` early only when `get_shallow_script` returns a null `Ref`, indicating that the script simply failed to load. Actual parsing errors are ignored at that point so that `reload` will always be called on a script that loads. Then parsing errors will not be silently lost. Note that `get_shallow_script` will not have called `GDScriptCompiler::make_scripts` in the case of a parsing error, but this is fine because `get_shallow_script` is only called here for scripts that are not in the `full_gdscript_cache` and therefore were never compiled, and `reload` will exit after finding the same error during its own call of `GDScriptParser::parse`, well before compilation is attempted. 

Other callers of `get_shallow_script` of course still retain the ability to error out if that initial parsing fails.

Fixes https://github.com/godotengine/godot/issues/75545.

### CODE REVIEW

File: modules/gdscript/gdscript_cache.cpp [hunk 1]
```diff
@@ -287,7 +287,8 @@ Ref<GDScript> GDScriptCache::get_full_script(const String &p_path, Error &r_erro
 
 	if (script.is_null()) {
 		script = get_shallow_script(p_path, r_error);
-		if (r_error) {
+		// Only exit early if script failed to load, otherwise let reload report errors
```
Comment (adamscott) [inline modules/gdscript/gdscript_cache.cpp:290]:
```suggestion
		// Only exit early if script failed to load, otherwise let reload report errors.
```

**Comment by adamscott** [comment]
Hi @mattbork. Thanks for your interest in Godot and your first PR!

Unfortunately, I fail to see any difference between your fix and the master branch. Can you join a MRP where I can easily test the difference? I have trouble running the MRP from #75545.

**Comment by mattbork** [comment]
Sorry, my mistake. I should have included the MRP I was testing against. The example in #75545 actually does report errors and break at runtime because it's not a syntax error but simply an identifier not in scope. A better example is to run the following script from the command line  with `--script`:

```
extends SceneTree

func _initialize()
        print("hello")
```

On master, the output is

```
ERROR: Failed to load script "res://cmd-mrp.gd" with error "Parse error".
   at: ResourceFormatLoaderGDScript::load (modules\gdscript\gdscript.cpp:2663)
```

Before #78540, there would have been no error reported at all. With the change here, the output now properly includes the actual parse error:

```
SCRIPT ERROR: Parse Error: Expected ":" after function declaration.
          at: GDScript::reload (res://cmd-mrp.gd:3)
ERROR: Failed to load script "res://cmd-mrp.gd" with error "Parse error".
   at: ResourceFormatLoaderGDScript::load (modules\gdscript\gdscript.cpp:2663)
```

As for in the editor, here's a project with a single node with a script attached.

[mrp.zip](https://github.com/godotengine/godot/files/13034410/mrp.zip)

The script has a simple syntax error:

```
extends Node

func _ready()
	print("hello")
```

On master, the editor notes the syntax error and the separate error from `ResourceFormatLoaderGDScript::load` is printed to the debugger's Error panel but the scene still runs:

no-fix

With these changes, the scene is interrupted by the debugger on line 3 and the actual syntax error is shown in the Stack Trace panel:

fix

**Review by adamscott — APPROVED** [review]
I tested the new MRP and it works as expected! Thanks for your contribution, it should be merged for the 4.2 release, as well as cherry-picked for other 4.0 and 4.1 releases.

Thanks! And don't hesitate @mattbork to join [the developer chat](https://chat.godotengine.org/). We're always welcoming new contributors.

**Comment by akien-mga** [comment]
Thanks! And congrats for your first merged Godot contribution :tada:

**Comment by YuriSizov** [comment]
Cherry-picked for 4.1.3.

**Pull request merged**